### PR TITLE
Fix object instantiation bug

### LINKED ISSUE #3: network arg not working on windows
## Abstract

On windows network arg is not working (It works on Linux though):

```bash
C:\Users\vagrant\Desktop\New folder>keytool network
keytool: An internal error occurred.

C:\Users\vagrant\Desktop\New folder>keytool asset
btc
```

Runtime error happens here:

```cpp
int main( int argc, char *argv[] ) {
    try {
        auto p = Params::parse(argc, argv);
        p->process();

        if(p->output.empty()) {
            throw runtime_error("An internal error occurred.");
        }
```

Presumably, the problem happens in `model()` constructor where the `network` is passed from the `asset`.

For unit tests to work the network option must be passed explicitly:

```bash
C:\Users\vagrant\Desktop\New folder>keytool --seed 8935a8068526d84da555cdb741a3b8a8 master-key
Assertion failed: false, file wally-utils.cpp, line 73

C:\Users\vagrant\Desktop\New folder>keytool --seed 8935a8068526d84da555cdb741a3b8a8 master-key --network testnet
tprv8ZgxMBicQKsPefYDL9LGdrXdW61RgXdc6Rshr7YBDzH39em5ALNVv8BG3bQBEXZGDPBe1Kkd86aqTuSj85iiAtMkDi8XeLaafsUdzcT9bn6
```

### PR DESCRIPTION
**Pull request by gorazdko**
## Abstract

Fixes #3 

The bug happened due to "race conditions" in object instantiation dependent on another object instantiation.

Linux builds and runs unit tests fine.

Windows builds and runs unit tests fine (ran manually)

## Status

Ready for review

### CODE REVIEW

File: src/network.cpp [hunk 1]
```diff
@@ -1,8 +1,5 @@
 #include "asset.hpp"
 
-Network Network::mainnet {"mainnet"};
-Network Network::testnet {"testnet"};
-
 std::vector<Network> Network::_networks = {
     mainnet, testnet
```
Comment (gorazdko) [inline src/network.cpp:4]: Hmm, now this line doesn't look right. When these elements are passed by value the same problem may occur here at some point.

**Comment by wolfmcnally** [comment]
What is the purpose in the PR for the changes to the `argp-standalone` submodule?

**Comment by wolfmcnally** [comment]
Is there documentation on these "race conditions" or best practices related thereto? This problem doesn't happen on Mac (or Linux either?) so it appears to be specific to Windows. Is it a compiler bug we're working around? If so, I'd like that to be very clear in the code, because to just make this change without explanation in the code (or conditional compilation to only compile it this way under Windows) is bad style.

**Comment by gorazdko** [comment]
> This problem doesn't happen on Mac (or Linux either?) so it appears to be specific to Windows. Is it a compiler bug we're working around?

It's logic bug. The fact that it runs on Linux and Mac is a happy accident. `Rust` compiler would caught this bug at compile time:

network.cpp:
```cpp
Network Network::mainnet {"mainnet"};
Network Network::testnet {"testnet"};
```
asset.cpp:
```cpp
Asset Asset::btc = {"Bitcoin", "btc", Network::mainnet, 0, 0x00, 0x05};
Asset Asset::btct = {"Bitcoin Testnet", "btct", Network::testnet, 1, 0x6f, 0xc4};
```
How do we know that `Network::mainnet` constructor has been executed before `Asset Asset::btc = {"Bitcoin", "btc", Network::mainnet, 0, 0x00, 0x05};`? We don't.

Object instantiation dependent on another object instantiation at "compile time" is bad practice. It should be put into initialization function.

> What is the purpose in the PR for the changes to the argp-standalone submodule?

The same as in `bc-seedtool-cli` - compiling on windows. It's not a change, it's a submodule update.

**Comment by gorazdko** [comment]
To point out the bug from another perspective:

Checkout the current master and move the two lines from `network.cpp` to `model.cpp`:

```cpp
Network Network::mainnet {"mainnet"};
Network Network::testnet {"testnet"};
```

The code builds on all platforms, but now `keytool network` fails on Linux (and possibly on macos as well). Works on Windows though.

**Comment by wolfmcnally** [comment]
Fair enough, but then shouldn't your fix be to use an initialization function as you suggest above, rather than simply move unrelated functions from one translation unit to another? That's also practice.

**Comment by gorazdko** [comment]
Fixed. Let me know if you find this fix better, or if you would do it differently.

I did answer the question wrt argp submodule:
> The same as in bc-seedtool-cli - compiling on windows. It's not a change, it's a submodule update to its latest state.

**Comment by wolfmcnally** [comment]
@gorazdko After further research, I determined that the best practice here is the *Construct members on first use idiom* as described [here](https://isocpp.org/wiki/faq/ctors#static-init-order-on-first-use-members). I implemented that in [this](https://github.com/BlockchainCommons/bc-keytool-cli/commit/3e313791b51b3e3919b46d2834cce4f6a4769cc7) commit. One advantage of this approach is not having to specifically call an initialization method at some early point in the program, eliminating boilerplate and another possible source of error. This approach also handles an arbitrarily complex directed graph of static initializations, and only initializes object actually referenced.

I also updated argp-standalone to the latest version in [this](https://github.com/BlockchainCommons/bc-keytool-cli/commit/33809da570c3237506ede5a5106ad7af578c3ed1) commit.

Please test under Linux and Windows and let me know whether it works.

**Comment by gorazdko** [comment]
ACK 0db752a7f8bdb61c159f4f0202807db4a9d51fcd

builds and unit tests pass on Linux and Windows.

**Pull request closed without merging**